### ISSUE TITLE
Conflict between nearest_neighbor() function and euclidean distance for agent pair.  

### ISSUE THREAD

**Issue by sahil-khan11** [JuliaDynamics/Agents.jl#1043]
**Issue**
The nearest neighbor for a given agent within the radius "r" doesn't align with the Euclidean distance between the agent and its nearest neighbor. The following example shows that the distance between neighbors is quite big compared to r used in the nearest_neighbor() function.  

**Minimal Working Example**
```using Agents, Random

@agent struct SocialAgent(ContinuousAgent{2, Float64})
    mass::Float64
end

function ball_model(; speed = 0.002)
    space2d = ContinuousSpace((1, 1); spacing = 0.02)
    model = StandardABM(SocialAgent, space2d; agent_step!, properties = Dict(:dt => 1.0),
                        rng = MersenneTwister(42))
    # And add some agents to the model
    for ind in 1:500
        pos = Tuple(rand(abmrng(model), 2))
        vel = sincos(2π * rand(abmrng(model))) .* speed
        add_agent!(pos, model, vel, 1.0)
    end
    return model
end

agent_step!(agent, model) = move_agent!(agent, model, model.dt)

function model_step!(model)
    for (a1, a2) in interacting_pairs(model, 0.01, :nearest)
        elastic_collision!(a1, a2, :mass)
    end
end

model = ball_model()

#### Checking euclidean distance for nearest pairs. 
r = 10^-100
euclidean_distance(model[3].pos, nearest_neighbor(model[3], model, r).pos, model)
```
*Output*
```
0.02637710352142322
```
**Agents v6.0.12**
**Julia v1.10.1**

**Comment by sahil-khan11** [comment]
What do you think about this? I was thinking maybe replacing nearest_neighbor() in interaction_pair() with this logic. 

```
r = 0.7
agent = model[27]
partner  = model[27]
partner_dist = 10.0.          # some random big number, just to create a variable so we can replace 
                                           # new partners distance after comparison. 
z = partner_dist                # For return verification 

for id in allids(model)

    if id != agent.id
        x = euclidean_distance(agent.pos, model[id].pos, model)

        if x <= r && x <= partner_dist
            partner_dist = x
            partner = model[id]
        end
    end

    if partner_dist == z
        return nothing
    else
        return (partner, partner_dist)
    end
  
end
```

It doesn't look elegant at the moment but I think it does the work :) 

Some outputs here: 

with r = 1
```
(SocialAgent(56, [0.4123437422626348, 0.8654172979541885], [0.0014994958181429924, -0.0013234471245092032], 1.0), 0.11955423401462202)
```

with  r = 0.7

```
(SocialAgent(56, [0.4123437422626348, 0.8654172979541885], [0.0014994958181429924, -0.0013234471245092032], 1.0), 0.11955423401462202)
```
with r = 0.5

```
(SocialAgent(56, [0.4123437422626348, 0.8654172979541885], [0.0014994958181429924, -0.0013234471245092032], 1.0), 0.11955423401462202)
```

with r = 0.1 

```
nothing
```

with r = 10^-100 

```
nothing
```

Let me know what you think :)

**Comment by Tortar** [comment]
I'm not on my computer to test this, but I think this can be fixed simply by using `nearby_ids_exact` instead of `nearby_ids` in `nearest_neighbors`

**Comment by sahil-khan11** [comment]
Yes,  it seems nice. I just checked but there is a difference that `nearby_ids()` doesn't count self-agent which is the point of reference while `nearby_ids_exact()` does. How it can be avoided? because it will be an issue otherwise.

**Comment by Tortar** [comment]
I don't think that it is true, `nearby_ids_exact` just filters out `nearby_ids`

**Comment by Tortar** [comment]
Fixed in #1044
